### DISCUSSION TITLE
Deploying Angular project to Vercel

### DISCUSSION THREAD

**Discussion by caiobukvar** [vercel/community#3554] [Help] [answered]
### Summary

How can I use environment variables from vercel within Angular?? I've tried for several hours and can't seem to find a fix. I have the environments folder generated in angular but, since they are not pushed to github, vercel can't seem to access the env variables. If I do pass env variables on deployment options @ vercel, I have no idea on how to access them via code, so I simply can't make an api call that needs a secret key to operate.

### Steps to Reproduce

---- **@environments/environment:**

```
export const environment = {
  production: true,
  RAPID_KEY: "mykeyhere" 
};
```

---- **@service created to make the http request:**

```
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Planet } from '../Planet';
import { ActivatedRoute } from '@angular/router';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class GetPlanetInfoService {
  private apiUrl: string;
  currentEnvironment !: string
  rapidKey: string

  constructor(private http: HttpClient, private route: ActivatedRoute) {
    this.apiUrl = `https://planets-info-by-newbapi.p.rapidapi.com/api/v1/planets`;
    this.rapidKey = environment.RAPID_KEY;  
  }

  getPlanetById(id: string): Observable<Planet> {
    const headers = new HttpHeaders({
      'X-RapidAPI-Key': this.rapidKey,
      'X-RapidAPI-Host': 'planets-info-by-newbapi.p.rapidapi.com'
    });

    const searchUrl = `${this.apiUrl}/${id}`;

    return this.http.get<Planet>(searchUrl, { headers });
  }
}

```

@this.rapidKey = environment.RAPID_KEY;   -> this does not work when deploying @ Vercel.

**Comment by amyegan** [comment]
Hey @caiobukvar. I wasn't able to replicate this problem in my test app based on the [Angular Boilerplate](https://vercel.com/templates/angular/angular-boilerplate) template. If you share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example), that will give us a better view of what is going wrong.

**Reply by caiobukvar** [reply]
The problem itself is on the deploy stage, I don't know how can I explain other than showing my [github repo ](https://github.com/caiobukvar/project-planets). 

I can try an explain a little better. 
If I only use environment key RAPID_KEY='rapidapikey' it works just fine. The problem is, for deploying on Vercel, I can't explicitly pass the apikey or it's considered a security breach. For that, I need to store that key on  a variable, usally on .env. Since i'm new to Angular, I just don't know how to access .env variables (react = process.env.key_name). I somehow think it's a simple configuration issue that I can't find an answer for.

**Reply by caiobukvar** [reply]
Or use variables from Vercel deploy environment, but they are not imported from environments/environment.ts, then project just doesn't build because ```import { environment } from '../../environments/environment'; ``` and ```this.rapidKey = environment.RAPID_KEY;  ``` does not exist since environments folder isn't published to github.

**Reply by amyegan** [reply, marked as answer]
You can use a custom webpack configuration or a tool like [ngx-env](https://github.com/chihab/ngx-env#using-environment-variables) to make environment variables available with `process.env` in Angular apps.

**Reply by caiobukvar** [reply]
Ok, will try to add it, thanks!

**Reply by fasilv843** [reply]
Did it work? I also faced the same error when I tried to deploy my project on angular. So I used ngx-env to solve it. but it didn't went well. 
This is a sample code how I implemented it using thier documentation

**src/environments/environments.prod.ts**
`
 export const environments = { 
    production: true,
    baseUrl: import.meta.env.NG_APP_BASE_URL
 }
`

**src/env.d.ts**

`
  interface ImportMeta {
   readonly env: ImportMetaEnv
  }
`
`
interface ImportMetaEnv {
  readonly NODE_ENV: string
  readonly NG_APP_BASE_URL: string
}
`


**.env**

`
NG_APP_BASE_URL = https://example.com
`

**package.json**

`
  "dependencies": {
    "@ngx-env/builder": "^16.1.3",
    "dotenv": "^16.3.1",
    "ngx-env": "^0.1.0"
  }
`

When I run ng serve, angular compiled the code without any error. But on the browser console shows this error and webpage is blank

![Screenshot 2024-01-19 192657](https://github.com/vercel/community/assets/124400398/3575af5b-a2cc-460b-9bb1-6b0e752e08d3)

It says `import.meta.env` is undefined. what did I miss here, is there any modules that i have to import?

**Reply by caiobukvar** [reply]
Hey! To be honest, I had to change projects and never touched it again. If I ever go back to it and it works, I can post it down here!
